sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/CastOperatorConversion.java

```diff
@@ -92,10 +92,13 @@ public DruidExpression toDruidExpression(
                                               ? ExpressionType.fromColumnType(fromDruidType)
                                               : ExpressionType.fromColumnType(toDruidType);
 
-      if (fromExpressionType == null || toExpressionType == null) {
+      if (toExpressionType == null) {
         // We have no runtime type for these SQL types.
         return null;
       }
+      if (fromExpressionType == null) {
+        return DruidExpression.ofLiteral(toDruidType, DruidExpression.nullLiteral());
+      }
```

Comment:
this isn't really a cool change to make as is at least, `Calcites.getColumnTypeForRelDataType` returns null for any SQL type we are unable to convert to a native druid type, it doesn't necessarily mean that the type is a null literal. see #17011